zkevm-circuits/src/keccak_circuit/keccak_padding_multirows_2.rs

```diff
@@ -0,0 +1,637 @@
+use crate::{evm_circuit::util::constraint_builder::BaseConstraintBuilder, util::Expr};
+use eth_types::Field;
+use gadgets::util::{not, select};
+use halo2_proofs::{
+    circuit::{Layouter, Region, SimpleFloorPlanner},
+    plonk::{Advice, Circuit, Column, ConstraintSystem, Error, Fixed, Selector},
+    poly::Rotation,
+};
+use std::marker::PhantomData;
+
+const KECCAK_REGION_WIDTH: usize = 64;
+const KECCAK_REGION_WIDTH_IN_BYTES: usize = KECCAK_REGION_WIDTH / 8;
+const KECCAK_REGION_HEIGHT: u64 = 17;
+const KECCAK_RATE: usize = 1088;
+const KECCAK_RATE_IN_BYTES: usize = KECCAK_RATE / 8;
+
+/// KeccakPaddingConfig
+#[derive(Clone, Debug)]
+pub struct KeccakPaddingConfig<F> {
+    q_enable: Selector,
+    q_end: Column<Advice>,
+    is_first_row: Column<Fixed>,
+    is_last_row: Column<Fixed>,
+    curr_padding_sum: Column<Advice>,
+    d_bits: [Column<Advice>; KECCAK_REGION_WIDTH],
+    d_lens: [Column<Advice>; KECCAK_REGION_WIDTH / 8],
+    d_rlcs: [Column<Advice>; KECCAK_REGION_WIDTH / 8],
+    s_flags: [Column<Advice>; KECCAK_REGION_WIDTH / 8],
+    randomness: Column<Advice>,
+
+    _marker: PhantomData<F>,
+}
+
+#[derive(Debug)]
+pub(crate) struct KeccakPaddingBlock<F: Field> {
+    pub(crate) q_end: u64,
+    pub(crate) acc_rlc: F,
+    pub(crate) acc_len: u32,
+    pub(crate) block_rows: [KeccakPaddingBlockRow<F>; KECCAK_REGION_HEIGHT as usize],
+}
+
+#[derive(Debug)]
+pub(crate) struct KeccakPaddingBlockRow<F: Field> {
+    pub(crate) curr_padding_sum: u32,
+    pub(crate) d_bits: [u8; KECCAK_REGION_WIDTH],
+    pub(crate) d_lens: [u32; KECCAK_REGION_WIDTH_IN_BYTES],
+    pub(crate) d_rlcs: [F; KECCAK_REGION_WIDTH_IN_BYTES],
+    pub(crate) s_flags: [bool; KECCAK_REGION_WIDTH_IN_BYTES],
+}
+
+/// KeccakPaddingCircuit
+#[derive(Default)]
+pub struct KeccakPaddingCircuit<F: Field> {
+    inputs: Vec<KeccakPaddingBlock<F>>,
+    size: usize,
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> KeccakPaddingCircuit<F> {
+    fn r() -> F {
+        F::from(123456)
+    }
+}
+
+impl<F: Field> Circuit<F> for KeccakPaddingCircuit<F> {
+    type Config = KeccakPaddingConfig<F>;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+        KeccakPaddingConfig::configure(meta)
+    }
+
+    fn synthesize(&self, config: Self::Config, layouter: impl Layouter<F>) -> Result<(), Error> {
+        config.assign(
+            layouter,
+            self.size,
+            &self.inputs[0],
+            KeccakPaddingCircuit::r(),
+        )?;
+        Ok(())
+    }
+}
+
+impl<F: Field> KeccakPaddingConfig<F> {
+    pub(crate) fn configure(meta: &mut ConstraintSystem<F>) -> Self {
+        let q_enable = meta.selector();
+        let q_end = meta.advice_column();
+        let is_first_row = meta.fixed_column();
+        let is_last_row = meta.fixed_column();
```

Comment:
You could make create these columns using `meta.selector()` as well, not really much of a difference though, in some specific cases halo2 may be able to do some optimizations internally if they are known to be selectors.